### DISCUSSION TITLE
index with $or operator not working

### DISCUSSION THREAD

**Discussion by brifai** [apache/couchdb#2063] [Q&A] [answered]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
I am trying to create an index for a query with "$or" operator but the query plan does not recognize the index.
my question is, how to create a working index for the following query ?
<pre>
{
   "selector": {
      "type": "project",
      "$or": [
         {
            "pmId": "8b1cee44bf16ccced9163adee20008a8"
         },
         {
            "sponsor._id": "8b1cee44bf16ccced9163adee20008a8"
         },
         {
            "AccessMembers": {
               "$elemMatch": {
                  "$and": [
                     {
                        "id": "8b1cee44bf16ccced9163adee20008a8"
                     },
                     {
                        "ACL": {
                           "$in": [
                              "View",
                              "Edit"
                           ]
                        }
                     }
                  ]
               }
            }
         }
      ]
   }
}
</pre>

## Steps to Reproduce

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB Version used:
v. 2.1.1
* Browser name and version:
* Operating System and version:
Windows Server 2016 Datacenter

**Comment by wohali** [comment]
What is your Mango index definition?
What does the `/{db}/_explain' endpoint say when you use it?

I would think in this situation you'd need every single field you're referencing in your index: `pmId`, `sponsor._id`, and `ACL`.

**Comment by brifai** [comment]
thank you @wohali for the reply.
the index is as following:
<pre>
{
	"index": {
		"fields": [
			"type",
			"pmId",
			"sponsor._id",
			"AccessMembers.[].id",
			"AccessMembers.[].ACL"
		]
	},
	"name": "pm-projects-index",
	"type": "json"
}
</pre>
the "explain" result as following:
<pre>
{
 "dbname": "pm",
 "index": {
  "ddoc": null,
  "name": "_all_docs",
  "type": "special",
  "def": {
   "fields": [
    {
     "_id": "asc"
    }
   ]
  }
 },
 "selector": {
  "$and": [
   {
    "type": {
     "$eq": "project"
    }
   },
   {
    "$or": [
     {
      "pmId": {
       "$eq": "8b1cee44bf16ccced9163adee20008a8"
      }
     },
     {
      "sponsor._id": {
       "$eq": "8b1cee44bf16ccced9163adee20008a8"
      }
     },
     {
      "AccessMembers": {
       "$elemMatch": {
        "$and": [
         {
          "id": {
           "$eq": "8b1cee44bf16ccced9163adee20008a8"
          }
         },
         {
          "ACL": {
           "$in": [
            "View",
            "Edit"
           ]
          }
         }
        ]
       }
      }
     }
    ]
   }
  ]
 },
 "opts": {
  "use_index": [],
  "bookmark": "nil",
  "limit": 25,
  "skip": 0,
  "sort": {},
  "fields": "all_fields",
  "r": [
   49
  ],
  "conflicts": false,
  "stale": false,
  "update": true,
  "stable": false,
  "execution_stats": false
 },
 "limit": 25,
 "skip": 0,
 "fields": "all_fields",
 "mrargs": {
  "include_docs": true,
  "view_type": "map",
  "reduce": false,
  "start_key": null,
  "end_key": "<MAX>",
  "direction": "fwd",
  "stable": false,
  "update": true,
  "conflicts": "undefined"
 }
}
</pre>

**Comment by garrensmith** [comment]
The simplest is just creating an index on type. The add in the execution stats to see how many docs are ignored or not.


________________________________

**Comment by brifai** [comment]
Thank you @garrensmith for your input, but can this query be indexed ?

**Comment by garrensmith** [comment, marked as answer]
If you use an $or operator then all the fields in the index need to be on both sides of the or otherwise it won’t work. That’s why I think keep the index fields to a minimum or use a partial_filter_selector when building the index.


________________________________

**Comment by brifai** [comment]
Thank you @garrensmith , @wohali do you have any input about it ?

**Comment by wohali** [comment]
no, garren is the expert here
